File: tests/js/quicksort/test_benchmark.mjs
import { quick_sort } from '../../../src/js/sorting/quicksort.mjs';
import { printArray } from '../test_utils.mjs';
import { isSorted } from '../test_utils.mjs';
import { arrayChecksum } from '../../../src/js/utils/utils.mjs';
import { loadSortData } from '../test_loaders.mjs';

const size = process.argv[2] || 'small';
const path = `../../../datasets/benchmark/sorting/${size}.bin`;
const sortData = loadSortData(inputPath);

const n = sortData.n;
const original = sortData.arr.slice();
const sorted = sortData.arr.slice();

quick_sort(sorted, sorted.length);

console.log(`Output (${n} elements):`);
console.log(printArray(sorted, 3));

const arraySorted = isSorted(sorted);
const sumIn = arrayChecksum(original);
const sumOut = arrayChecksum(sorted);
const sameSum = sumIn === sumOut;

console.log('RESULT', JSON.stringify({
    n,
    sorted: arraySorted,
    same_sum: sameSum,
    first: sorted[0],
    last: sorted[n - 1],
    checksum: sumOut.toString()
}));

File: tests/c/quicksort/test_benchmark.c
/* Quick sort testing with benchmark datasets */
#include <stdio.h>
#include <stdlib.h>
#include "../../../src/c/utils/utils.h"

void quick_sort(int *arr, int n);

int main(int argc, char *argv[]) {
    const char *pathType = "small"; // default to small

    if (argc >= 2) pathType = argv[1];

    char path[256];
    snprintf(path, sizeof(path), "../../../datasets/benchmark/sorting/%s.bin", pathType);

    int n;
    int *original = load_sort_data(path, &n);
    int *arr = copy_array(original, n);

    printf("Quick sort\n");
    printf("Dataset: %s\n", path);
    printf("Input (%d elements): ", n);
    print_array(original, n, 3);

    quick_sort(arr, n);

    printf("Output (%d elements): ", n);
    print_array(arr, n, 3);

    int sorted = is_sorted(arr, n);
    long long sum_in = array_checksum(original, n);
    long long sum_out = array_checksum(arr, n);
    int same_sum = (sum_in == sum_out);

    printf("RESULT {\"n\":%d,\"sorted\":%s,\"same_sum\":%s,\"first\":%d,\"last\":%d,\"checksum\":\"%lld\"}\n", n, sorted ? "true" : "false", same_sum ? "true" : "false", arr[0], arr[n - 1], sum_out);

    free(original);
    free(arr);
    return 0;
}